### PR TITLE
Bug Fixes on start script for rest api jetty.

### PR DESCRIPTION
**Pull request by wjonassen**
Updates to get each sub project building and running properly, except for web-client-jetty.

## Problem Description

Fixes #. 

Describe the problem you are trying to solve.

## Solution

Describe your solution.

## how you tested the change

Describe what was done to test the change. This section can be left blank 
if automated tests demonstrating usage are provided in the PR.

## Where the following done:

- [ ] Tests. Check all that apply:
   - [ ] Unit tests created or modified that run during ant test.
   - [ ] Integration tests created or modified that run during integration testing
         (Formerly called regression tests.)
   - [ ] Test procedure descriptions for manual testing
- [ ] Was relevant documentation updated?
- [ ] Were relevant config element (e.g. XML data) updated as appropriate

If you aren't sure leave unchecked and we will help guide you to want needs changing where.

### CODE REVIEW

File: gradle/libs.versions.toml [hunk 1]
```diff
@@ -1,13 +1,13 @@
 [versions]
-opendcs = "7.0.10"
+opendcs = "7.0.11"
 
 jetty = "9.4.53.v20231009"
-servlet-api = "3.1.0"
-slf4j = "2.0.0"
+servlet-api = "4.0.1"
```
Comment (MikeNeilson) [inline gradle/libs.versions.toml:5]: This shouldn't be updated arbitrarily. certain versions require jakarta.* namespace and java 11+

Is there a reason this was updated?

File: opendcs-rest-api-jetty/build.gradle [hunk 2]
```diff
@@ -46,15 +44,19 @@ task startJetty(type: JavaExec) { ->
...
-task prepareStartScript(type: Copy) {
-    from './bin/start.sh'
-    into "${buildDir}/odcsapi-jetty-standalone/bin/"
-    filter(ReplaceTokens, tokens: ['jarName': jar.archiveFileName.get()])
+task createStartScript() {
+    doLast {
+        new File("${buildDir}/bin", "start.sh").text = """
+#!/bin/bash
+cd `dirname \$0`/..
+java \$JAVA_OPTS -jar libs/${project.name}-${project.version}.jar \$JAVA_ARGS
```
Comment (MikeNeilson) [inline opendcs-rest-api-jetty/build.gradle:52]: There appears to be an issue in the chain of tasks. The action build is showing start.sh as missing.

File: opendcs-rest-api-jetty/src/main/java/org/opendcs/odcsapi/jetty/Start.java [hunk 3]
```diff
@@ -107,7 +105,8 @@ public static void main(String[] args)
 		appLogger.info("================ Starting ===============");
 
 		// Initialize the JETTY server and servlet holders.
-		Server server = new Server();
+		org.eclipse.jetty.server.Server server = new org.eclipse.jetty.server.Server();
```
Comment (MikeNeilson) [inline opendcs-rest-api-jetty/src/main/java/org/opendcs/odcsapi/jetty/Start.java:108]: 1. Why?
2. just delete comment code, don't leave it in.

**Review by MikeNeilson — CHANGES_REQUESTED** [review]

File: opendcs-rest-api-jetty/build.gradle (hunk 2 above)
Comment (wjonassen) [reply opendcs-rest-api-jetty/build.gradle:52]: It's fixed on my end, going to do a new push soon.

File: opendcs-rest-api-jetty/src/main/java/org/opendcs/odcsapi/jetty/Start.java (hunk 3 above)
Comment (wjonassen) [reply opendcs-rest-api-jetty/src/main/java/org/opendcs/odcsapi/jetty/Start.java:108]: Adam and I were just going through the process of pushing these changes in git via IntelliJ.

I am cleaning it up and will be pushing the cleaned code soon.

Comment (MikeNeilson) [reply opendcs-rest-api-jetty/src/main/java/org/opendcs/odcsapi/jetty/Start.java:108]: ah, okay.

File: gradle/libs.versions.toml (hunk 1 above)
Comment (wjonassen) [reply gradle/libs.versions.toml:5]: Good point, I will revert this.  I left this pull request in draft because I knew there were some items that needed some cleaning up..  We were just doing it through intellij so I could get the hang of doing it through intellij.  I will clean it up and when it's ready make it a non-draft.

File: opendcs-rest-api-jetty/build.gradle [hunk 4]
```diff
@@ -46,15 +44,26 @@ task startJetty(type: JavaExec) { ->
...
 
-task prepareStartScript(type: Copy) {
-    from './bin/start.sh'
-    into "${buildDir}/odcsapi-jetty-standalone/bin/"
-    filter(ReplaceTokens, tokens: ['jarName': jar.archiveFileName.get()])
+task createStartScript() {
+    doLast {
+
+        //Make bin directory if it does not exist already.  buildDir is now deprecated.
+        File buildDir = project.layout.getBuildDirectory().get().getAsFile();
```
Comment (wjonassen) [inline opendcs-rest-api-jetty/build.gradle:51]: There is probably a better way to do this.  project.buildDir was deprecated, so this was one way to get it done.  Looking into a less convoluted way.

Comment (MikeNeilson) [reply opendcs-rest-api-jetty/build.gradle:51]: Honestly it's probably fine. there are ways to properly define inputs and outputs from tasks but in this context that's probably going to look even more convoluted than what you have..

Comment (MikeNeilson) [reply opendcs-rest-api-jetty/build.gradle:51]: Hmm, actually maybe not. This is similar is concept: https://github.com/USACE/cwms-data-api/blob/95735d01d6ddada96b8eb8aa53a07b392197548f/cwms-data-api/build.gradle#L232

You might consider this style.

Comment (wjonassen) [reply opendcs-rest-api-jetty/build.gradle:51]: Looks like this one is using "$buildDir" (A much simpler approach) - which was deprecated in 8.5.
https://github.com/gradle/gradle/issues/20210

Comment (MikeNeilson) [reply opendcs-rest-api-jetty/build.gradle:51]: we really need to update that particular build. still gradle 6.4

File: gradle/libs.versions.toml (hunk 1 above)
Comment (adamkorynta) [reply gradle/libs.versions.toml]: 4.0.1 was the final release before migrating to jakarta. The maven artifact shows it was compiled with JDK 8: https://mvnrepository.com/artifact/javax.servlet/javax.servlet-api/4.0.1

**Comment by rma-psmorris** [comment]
https://github.com/opendcs/rest_api/issues/38

File: opendcs-rest-api-jetty/src/main/java/org/opendcs/odcsapi/jetty/Start.java [hunk 5]
```diff
@@ -73,41 +61,38 @@ public static void main(String[] args)
...
+			is = Start.class.getClassLoader().
+					getResourceAsStream("logging.properties");
+		}
+		else
+		{
+			System.out.println("Logging File Passed: " + apiCmdLineArgs.getLoggingPropertiesFile());
+			File lf = new File(loggingFile);
+			is = new FileInputStream(lf);
+		}
+		LogManager.getLogManager().readConfiguration(is);
```
Comment (wjonassen) [inline opendcs-rest-api-jetty/src/main/java/org/opendcs/odcsapi/jetty/Start.java:79]: This was flagged as a possible security issue.  I would like to give the user the ability to pass the logging.properties file to the start script if they don't want to use the default one in the resources folder.

File: opendcs-web-client-jetty/src/main/java/org/opendcs/odcswebclient/jetty/Start.java [hunk 6]
```diff
@@ -37,30 +38,57 @@ public class Start
...
             System.exit(1);
         }
+
+        try
+        {
+            apiFileDetailsPath = argParse.switchValue("-f",  null);
+        }
+        catch (Exception e)
+        {
+            System.out.println("No Api File Details Path File Provided.  Using defaults.");
```
Comment (wjonassen) [inline opendcs-web-client-jetty/src/main/java/org/opendcs/odcswebclient/jetty/Start.java:68]: For the system.outs in this file, a logger will be added, but that will be in a separate pull request.  the Start.java file in opendcs-rest-api-jetty is using a logger and I want to make sure that looks good before creating the new pull request for the rest of the project.

File: opendcs-rest-api/src/main/resources/logging.properties
Comment (adamkorynta) [inline opendcs-rest-api/src/main/resources/logging.properties:1]: This should be included in the jetty installer instead of the war/jar with the -D flag pointing.

Here's an example properties file distributed with other HEC applications:
[logging.properties.txt](https://github.com/opendcs/rest_api/files/13953621/logging.properties.txt)

File: opendcs-web-client-jetty/build.gradle [hunk 7]
```diff
@@ -4,12 +4,18 @@ plugins {
...
 dependencies {
-    implementation(project(":opendcs-web-client"))
+    implementation(libs.jetty.jsp)
     implementation(libs.jetty.server)
+    implementation(libs.javax.servlet.jsp)
+    implementation(libs.servlet.api)
+    implementation(libs.jetty.annotations)
     implementation(libs.jetty.webapp)
+    implementation(libs.jersey.container.servlet)
+    implementation(libs.tomcat.embed)
```
Comment (adamkorynta) [inline opendcs-web-client-jetty/build.gradle:18]: Tomcat dependency in the jetty installer?

File: opendcs-web-client-jetty/build.gradle [hunk 8]
```diff
@@ -35,15 +41,51 @@ task startJetty(type: JavaExec) { ->
...
+port=8080
+context=odcsapi
+"""
+    }
+}
+
+task filepermission() {
+    doLast {
+        String start_path = "${buildDir}\\bin\\start.sh"
+        System.out.println("Here: " + start_path)
```
Comment (adamkorynta) [inline opendcs-web-client-jetty/build.gradle:78]: Remove debug statement (or use the gradle logger)

File: opendcs-web-client/build.gradle [hunk 9]
```diff
@@ -6,7 +6,7 @@ plugins {
 }
 
 dependencies {
-    implementation(libs.servlet.api)
+    implementation(libs.jetty.jsp)
```
Comment (adamkorynta) [inline opendcs-web-client/build.gradle:9]: A jetty dependency here is suspicious given this should also run in the context of tomcat

File: opendcs-web-client/src/main/java/api/Gateway.java [hunk 10]
```diff
@@ -124,16 +124,27 @@ protected void doDelete(HttpServletRequest request, HttpServletResponse response
      */
     public void setApiDetails(ServletContext sc) throws IOException
     {
-        //Gets the initialization details from the conf file.
-        String apiConfigFilePath;
-        if (sc == null)
-        {
-            apiConfigFilePath = this.getServletContext().getRealPath("/conf/api.conf");
+        //this.baseUrl = "http://108.20.179.32:9876";
```
Comment (adamkorynta) [inline opendcs-web-client/src/main/java/api/Gateway.java:127]: Dead code

File: opendcs-web-client/src/main/java/api/Gateway.java [hunk 11]
```diff
@@ -124,16 +124,27 @@ protected void doDelete(HttpServletRequest request, HttpServletResponse response
...
         }
-        else
+        Object configAttr = sc.getAttribute("api_details_file_path");
+        String apiConfigFilePath = (configAttr == null) ? null : configAttr.toString();
+
+        if (apiConfigFilePath == null)
         {
+            //Use the config file within WAR file.
+            //default api file location.
+            //if (sc == null) {
```
Comment (adamkorynta) [inline opendcs-web-client/src/main/java/api/Gateway.java:139]: Dead code

File: opendcs-web-client-jetty/src/main/java/org/opendcs/odcswebclient/jetty/Start.java [hunk 12]
```diff
@@ -37,30 +38,56 @@ public class Start
     public static void main(String[] args) 
             throws Exception
     {
+        Log.setLog(new StdErrLog());
```
Comment (adamkorynta) [inline opendcs-web-client-jetty/src/main/java/org/opendcs/odcswebclient/jetty/Start.java:41]: How does this jetty log work in conjunction with slf4j/jul logging?

File: opendcs-rest-api/src/main/resources/swaggerui/open_api.json [hunk 13]
```diff
@@ -6,6 +6,12 @@
     "version": "0.0.3"
   },
   "servers": [
+    {
+      "url": ""
+    },
+    {
+      "url": "/opendcs-rest-api-99.bugfix-lib-updates-SNAPSHOT"
```
Comment (adamkorynta) [inline opendcs-rest-api/src/main/resources/swaggerui/open_api.json:13]: Remove debug code

File: opendcs-web-client-jetty/build.gradle (hunk 7 above)
Comment (wjonassen) [reply opendcs-web-client-jetty/build.gradle:18]: When I don't have this library added, I receive the following error when starting jetty:

java.lang.ClassNotFoundException: org.apache.catalina.filters.HttpHeaderSecurityFilter
        at java.net.URLClassLoader.findClass(URLClassLoader.java:387)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:418)
        at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:352)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:351)
        at org.eclipse.jetty.webapp.WebAppClassLoader.loadClass(WebAppClassLoader.java:538)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:351)
        at org.eclipse.jetty.util.Loader.loadClass(Loader.java:64)
        at org.eclipse.jetty.servlet.BaseHolder.doStart(BaseHolder.java:89)
        at org.eclipse.jetty.servlet.FilterHolder.doStart(FilterHolder.java:97)
        at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:73)
        at org.eclipse.jetty.servlet.ServletHandler.lambda$initialize$0(ServletHandler.java:749)
        at java.util.Spliterators$ArraySpliterator.forEachRemaining(Spliterators.java:948)
        at java.util.stream.Streams$ConcatSpliterator.forEachRemaining(Streams.java:742)
        at java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:647)
        at org.eclipse.jetty.servlet.ServletHandler.initialize(ServletHandler.java:774)
        at org.eclipse.jetty.servlet.ServletContextHandler.startContext(ServletContextHandler.java:379)
        at org.eclipse.jetty.webapp.WebAppContext.startWebapp(WebAppContext.java:1449)
        at org.eclipse.jetty.webapp.WebAppContext.startContext(WebAppContext.java:1414)
        at org.eclipse.jetty.server.handler.ContextHandler.doStart(ContextHandler.java:916)
        at org.eclipse.jetty.servlet.ServletContextHandler.doStart(ServletContextHandler.java:288)
        at org.eclipse.jetty.webapp.WebAppContext.doStart(WebAppContext.java:524)
        at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:73)
        at org.eclipse.jetty.util.component.ContainerLifeCycle.start(ContainerLifeCycle.java:169)
        at org.eclipse.jetty.server.Server.start(Server.java:423)
        at org.eclipse.jetty.util.component.ContainerLifeCycle.doStart(ContainerLifeCycle.java:110)
        at org.eclipse.jetty.server.handler.AbstractHandler.doStart(AbstractHandler.java:97)
        at org.eclipse.jetty.server.Server.doStart(Server.java:387)
        at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:73)
        at org.opendcs.odcswebclient.jetty.Start.main(Start.java:131)
2024-01-16 21:23:18.681:WARN:oejw.WebAppContext:main: Failed startup of context o.e.j.w.WebAppContext@31ef45e3{                                              OpenDCS Web Application,/odcsweb,file:///tmp/jetty-0_0_0_0-8083-opendcs-web-client-99_bugfix-lib-updates-SNAPSH                                              OT_war-_odcsweb-any-1401972714019795226/webapp/,UNAVAILABLE}{/home/opendcs/OPENDCS/odcs-web-client/bin/opendcs-                                              web-client-99.bugfix-lib-updates-SNAPSHOT.war}
javax.servlet.UnavailableException: Class loading error for holder httpHeaderSecurity==org.apache.catalina.filt                                              ers.HttpHeaderSecurityFilter@134593bf{inst=false,async=true,src=DESCRIPTOR:file:///tmp/jetty-0_0_0_0-8083-opend                                              cs-web-client-99_bugfix-lib-updates-SNAPSHOT_war-_odcsweb-any-1401972714019795226/webapp/WEB-INF/web.xml}
        at org.eclipse.jetty.servlet.BaseHolder.doStart(BaseHolder.java:96)
        at org.eclipse.jetty.servlet.FilterHolder.doStart(FilterHolder.java:97)
        at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:73)
        at org.eclipse.jetty.servlet.ServletHandler.lambda$initialize$0(ServletHandler.java:749)
        at java.util.Spliterators$ArraySpliterator.forEachRemaining(Spliterators.java:948)
        at java.util.stream.Streams$ConcatSpliterator.forEachRemaining(Streams.java:742)
        at java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:647)
        at org.eclipse.jetty.servlet.ServletHandler.initialize(ServletHandler.java:774)
        at org.eclipse.jetty.servlet.ServletContextHandler.startContext(ServletContextHandler.java:379)
        at org.eclipse.jetty.webapp.WebAppContext.startWebapp(WebAppContext.java:1449)
        at org.eclipse.jetty.webapp.WebAppContext.startContext(WebAppContext.java:1414)
        at org.eclipse.jetty.server.handler.ContextHandler.doStart(ContextHandler.java:916)
        at org.eclipse.jetty.servlet.ServletContextHandler.doStart(ServletContextHandler.java:288)
        at org.eclipse.jetty.webapp.WebAppContext.doStart(WebAppContext.java:524)
        at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:73)
        at org.eclipse.jetty.util.component.ContainerLifeCycle.start(ContainerLifeCycle.java:169)
        at org.eclipse.jetty.server.Server.start(Server.java:423)
        at org.eclipse.jetty.util.component.ContainerLifeCycle.doStart(ContainerLifeCycle.java:110)
        at org.eclipse.jetty.server.handler.AbstractHandler.doStart(AbstractHandler.java:97)
        at org.eclipse.jetty.server.Server.doStart(Server.java:387)
        at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:73)
        at org.opendcs.odcswebclient.jetty.Start.main(Start.java:131)

Comment (MikeNeilson) [reply opendcs-web-client-jetty/build.gradle:18]:
```
javax.servlet.UnavailableException: Class loading error for holder httpHeaderSecurity==org.apache.catalina.filt ers.HttpHeaderSecurityFilter@134593bf{inst=false,async=true,src=DESCRIPTOR:file:///tmp/jetty-0_0_0_0-8083-opend cs-web-client-99_bugfix-lib-updates-SNAPSHOT_war-_odcsweb-any-1401972714019795226/webapp/WEB-INF/web.xml}
```

Something is creating a really bad web.xml, there shouldn't be any references to a org.apache.catalina.filters class in a Jetty system. Something is very wrong, but it's not solved by including Tomcat embedded at this level.

Comment (adamkorynta) [reply opendcs-web-client-jetty/build.gradle:18]: That looks to be caused by the web.xml file: https://github.com/opendcs/rest_api/blob/main/opendcs-web-client/src/main/webapp/WEB-INF/web.xml#L184

File: opendcs-web-client-jetty/build.gradle (hunk 8 above)
Comment (wjonassen) [reply opendcs-web-client-jetty/build.gradle:78]: This function does not need to be in there.  Removing it.

File: opendcs-web-client/build.gradle (hunk 9 above)
Comment (wjonassen) [reply opendcs-web-client/build.gradle:9]: I created issue #92 to track this.

File: opendcs-web-client-jetty/build.gradle (hunk 7 above)
Comment (wjonassen) [reply opendcs-web-client-jetty/build.gradle:18]: I was able to clean up the WAR file and remove tomcat embedded.

File: opendcs-rest-api-jetty/build.gradle [hunk 14]
```diff
@@ -58,16 +58,94 @@ task createStartScript() {
         new File(binDir, "start.sh").text = """
 #!/bin/bash
 cd `dirname \$0`/..
-java \$JAVA_OPTS -jar libs/${project.name}-${project.version}.jar \$JAVA_ARGS
+java  -Djava.util.logging.config.file=logging.properties \$JAVA_OPTS -jar libs/${project.name}-${project.version}.jar \$JAVA_ARGS
             """
     }
 }
 
+task createLoggingPropertiesFile() {
```
Comment (adamkorynta) [inline opendcs-rest-api-jetty/build.gradle:66]: I think this one shouldn't get created in gradle, it should be a resource file that gets copied over

File: opendcs-rest-api-jetty/src/main/java/org/opendcs/odcsapi/jetty/Start.java [hunk 15]
```diff
@@ -166,31 +176,42 @@ else if ("siteNameTypePreference".equalsIgnoreCase(n))
...
 		}
 		ds.setUser(afr.getUsername());
 		ds.setPassword(afr.getPassword());
 		DbInterface.setDataSource(ds);
-		DbInterface.setDatabaseType(dbType);
+		try
+		{
+			DbInterface.setDatabaseType(dbType);
+		}
+		catch (Exception e)
```
Comment (adamkorynta) [inline opendcs-rest-api-jetty/src/main/java/org/opendcs/odcsapi/jetty/Start.java:191]: Do we want to continue if this throws an exception? Also if dbType is not null we are double logging

File: opendcs-rest-api-jetty/src/main/java/org/opendcs/odcsapi/jetty/Start.java [hunk 16]
```diff
@@ -166,31 +176,42 @@ else if ("siteNameTypePreference".equalsIgnoreCase(n))
 		{
 			afr.read();
 		}
-		catch(Exception ex)
+		catch(Exception e)
```
Comment (adamkorynta) [inline opendcs-rest-api-jetty/src/main/java/org/opendcs/odcsapi/jetty/Start.java:179]: The previous logic exited (by throwing an exception), but now it will continue.
